### ISSUE TITLE
Can we make it so the embedder doesn't need to call Browser::handle_event() when there's no event?

### ISSUE THREAD

**Issue by paulrouget** [servo/servo#15734]
_This issue is part of an effort to improve Servo's embedding story. See https://github.com/paulrouget/servoshell/projects/2_

By embedder, I mean what lives in `/ports/`.

If I'm not mistaken, libservo wakes up the main thread every time `compositor_proxy.send` is called ([example](https://github.com/servo/servo/blob/26de7c6bc48affbc2087b32649850f0733e567f0/ports/glutin/window.rs#L1122)). This is when the gl buffers are swapped (handled by the embedder). Now, it is apparently necessary after the main thread is awaken to call `Browser::handle_event(vec![])` (no event required), otherwise, nothing is rendered on the screen.

Can we make it so this is not necessary? To me it sounds like this can be handled internally.

**Comment by paulrouget** [comment]
To illustrate this problem, apply this patch:


```diff
diff --git a/ports/servo/main.rs b/ports/servo/main.rs
index b24e550ff2..f99e31ba9c 100644
--- a/ports/servo/main.rs
+++ b/ports/servo/main.rs
@@ -154,9 +154,12 @@ fn main() {
     // Feed events from the window to the browser until the browser
     // says to stop.
     loop {
-        let should_continue = browser.browser.handle_events(window.wait_events());
-        if !should_continue {
-            break;
+        let events = window.wait_events();
+        if !events.is_empty() {
+            let should_continue = browser.browser.handle_events(events);
+            if !should_continue {
+                break;
+            }
         }
     }
```

Servo won't render anything until the mouse (or any event get forwarded).

**Comment by metajack** [comment]
@asajeffrey Can you look into this?

**Comment by asajeffrey** [comment]
Will do.

**Comment by paulrouget** [comment]
Also, I'd like to understand what's the role of the Present event?

I'm embedding Servo, and I'm under the impression that I need to:
- swap gl buffers every time the main thread is awaken
- call Servo::handle_event() every time the main thread is awaken
- do nothing special when the `WindowMethods::present` is called

I would have imagined that the embedder only task was to swap the buffers when `WindowMethods::present` is called.

**Comment by asajeffrey** [comment]
@paulrouget this issue is next on my queue after x-origin window.parent. Is that okay, or are you blocked on this?

**Comment by paulrouget** [comment]
It's ok. I'm not blocked on this.

**Comment by asajeffrey** [comment]
OK, submitted my x-origin window.parent PR, move this to the front of the queue...

What behaviour would you prefer? Would you rather have the servo threads not block waiting for the app window? I can imagine that OpenGL puts restrictions that some gfx actions have to be performed by the main thread, so this might impact our ability to have the compositor perform actions independently.

@glennw am I correct that `Browser::handle_events` has to be called by the main thread?

**Comment by paulrouget** [comment]
I need clarifications and maybe a cosmetic change.

Clarifications:

I need to understand when exactly to swap the gl buffers. When `WindowMethods::Present` is called? Every time the main thread is awakened? Both? Should I swap after or before `handle_event()` has been called? Do I need to call `handle_event()` every time the main thread is awakened or only when it is awakened by the CompositorProxy?

Cosmetic change:

So even if there's no event to send to Servo, it's still necessary to call `Browser::handle_events(vec![])`. It's confusing because we would expect the browser to only handle the WindowEvents (the only parameter), where it actually handles messages from different sources. It took me some time to understand that I was supposed to called handle_events even though I had no events to send.

I was thinking maybe, within `handle_events`, WindowEvents would be collected from the window (new method: `WindowMethods::get_events() -> Vec<WindowEvent>`), and we could rename `fn handle_events(Vec<WindowEvent>)` `fn awakened()` or something like that.

I can take care of writing the PR. I just need to know if my reasoning is right.

**Comment by asajeffrey** [comment]
Clarifications: @glennw, you are better equipped than I am to talk about GL buffer swapping.

Cosmetic change: I'm not sure what the proposed API change is here, is the idea for the `Browser` to separate the method for queuing events from the one from performing updates?

**Comment by paulrouget** [comment]
> Cosmetic change: I'm not sure what the proposed API change is here, is the idea for the Browser to separate the method for queuing events from the one from performing updates?

Yes. It doesn't change the final behaviour, it just to make it clear what's going on.

The embedder would call `Browser::perform_updates()`, and in `perform_updates()`, Servo would call `window.get_events() -> Vec<WindowEvent>`.

**Comment by glennw** [comment]
```WindowMethods::present()``` is called by the compositor each time it has drawn a new frame to the backbuffer. So this callback is what should be used to swap the buffers.

**Comment by paulrouget** [comment]
Glenn helped me on IRC.

> I need to understand when exactly to swap the gl buffers. When WindowMethods::Present is called?

Yes.

> Every time the main thread is awakened? Both?

No.

> Should I swap after or before handle_event() has been called?

Before.

> Do I need to call handle_event() every time the main thread is awakened or only when it is awakened by the CompositorProxy?

Still unknown. Experimenting.

**Comment by asajeffrey** [comment]
@paulrouget: Ah, so the entry point from the embedder to servo would be a heartbeat `perform_updates()` and the API for events would turn from a push API to a pull API. This seems sensible, especially if we're trying to move to a more streams-based API.

**Comment by paulrouget** [comment]
@asajeffrey exactly!

**Comment by paulrouget** [comment]
Ideally, every time the compositor proxy wakes up the event loop, servo would call Servo::handle_events automatically, and pull the WindowEvents from somewhere. Then the embedder doesn't need to handle the special case of when there's no events to send to servo.

**Comment by mrobinson** [comment]
I think this is effectively done. Now when Servo wakes up the event loop, the embedder should just call `Servo::spin_event_loop` which will read any incoming messages and handle them. The embedder doesn't have to worry about how the internal messaging bits of Servo work.
